SSA/Projects/CIRCT/Comb/Comb.lean

```diff
@@ -0,0 +1,617 @@
+  | .extract w n => (Ty.bv w) → (Ty.bv (w - n))
+  | .icmp _ w => (Ty.bv w, Ty.bv w) → (Ty.bool)
+  | .mods w => (Ty.bv w, Ty.bv w) → (Ty.bv w)
+  | .modu w => (Ty.bv w, Ty.bv w) → (Ty.bv w)
+  | .mul w n => ${List.replicate n (Ty.bv w)} → (Ty.bv w)
+  | .mux w => (Ty.bv w, Ty.bv w, Ty.bool) → (Ty.bv w)
+  | .or w n => ${List.replicate n (Ty.bv w)} → (Ty.bv w)
+  | .parity w => (Ty.bv w) → (Ty.bool)
+  | .replicate w n => (Ty.bv w) → (Ty.bv (w * n))
+  | .shl w => (Ty.bv w, Ty.bv w) → (Ty.bv w)
+  | .shrs w => (Ty.bv w, Ty.bv w) → (Ty.bv w)
+  | .shru w => (Ty.bv w, Ty.bv w) → (Ty.bv w)
+  | .sub w => (Ty.bv w, Ty.bv w) → (Ty.bv w)
+  | .xor w n => ${List.replicate n (Ty.bv w)} → (Ty.bv w)
+
+instance : TyDenote (Dialect.Ty Comb) where
+  toType := fun
+  | .bv w => BitVec w
+  | .nat  => Nat
+  | .bool => Bool
+  | .hList l => HVector BitVec l -- het list of bitvec whose lengths are contained in l
+  | .icmpPred => CombOp.IcmpPredicate
+
+def HVector.replicateToList {α : Type} {f : α → Type} {a : α} :
+    {n : Nat} → HVector f (List.replicate n a) → List (f a)
+  | 0, _ => []
+  | n+1, HVector.cons x xs => x :: replicateToList xs
+
+def ofString? (s : String) : Option CombOp.IcmpPredicate :=
+  match s with
+  | "eq" => some .eq
+  | "ne" => some .ne
+  | "slt" => some .slt
+  | "sle" => some .sle
+  | "sgt" => some .sgt
+  | "sge" => some .sge
+  | "ult" => some .ult
+  | "ule" => some .ule
+  | "ugt" => some .ugt
+  | "uge" => some .uge
+  | _     => none
+
+def_denote for Comb where
+  | .add _ _ => fun xs => CombOp.add (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .and _ _ => fun xs => CombOp.and (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .concat _ => fun xs => CombOp.concat xs
+  | .divs _ => fun xs => CombOp.divs xs
+  | .divu _ => fun xs => CombOp.divu xs
+  | .extract _ n => fun xs => CombOp.extract xs n
+  | .icmp p _ => fun xs =>
+      let p' := Option.get! (ofString? p)
+      CombOp.icmp p' xs
+  | .mods _ => fun xs => CombOp.mods xs
+  | .modu _ => fun xs => CombOp.modu xs
+  | .mul _ _ => fun xs => CombOp.mul (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .mux _ => fun xs => CombOp.mux xs
+  | .or _ _ => fun xs => CombOp.or (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .parity _ => fun xs => CombOp.parity xs
+  | .replicate _ n => fun xs => CombOp.replicate xs n
+  | .shl _ => fun xs => CombOp.shl xs
+  | .shrs _ => fun xs => CombOp.shrs xs
+  | .shru _ => fun xs => CombOp.shru xs
+  | .sub _ => fun xs => CombOp.sub xs
+  | .xor _ _ => fun xs => CombOp.xor (HVector.replicateToList (f := TyDenote.toType) xs)
+
+end Dialect
+
+/-- we currently do not support the parsing of heterogeneous vectors -/
+def mkTy : MLIR.AST.MLIRType φ → MLIR.AST.ExceptM Comb Ty
+  | MLIR.AST.MLIRType.undefined s => do
+    match s.splitOn "_" with
+    | ["Bool"] =>
+      return .bool
+    | ["Nat"] =>
+      return .nat
+    | ["BitVec", r] =>
+      return .bv (r.toNat!)
```

Comment:
You might want to match on `MLIRType.int` instead, and map that to your `.bv` type (have a look at LLVM, I believe that's what we do there as well).

That way you can use the builtin `i64` type in your programs (which I believe CIRCT uses for comb, right?), and also you won't have to use `String.splitOn` here anymore, you can just directly match on the strings for the other cases.

Reply:
[`comb`](https://circt.llvm.org/docs/Dialects/Comb/#combshl-circtcombshlop) actually uses bitvecs for almost everything afaict, so it would be nice to maintain that I guess?

Reply:
Yes, "signless integer bitvector" in the documentation means `i$x`, for some `$x`. See, e.g., the [canonicalization](https://github.com/llvm/circt/blob/main/test/Dialect/Comb/canonicalization.mlir) test file which has the following example program:
```
hw.module @idempotentDeduped1(in %arg0 : i7, in %arg1 : i7, out resAnd: i7, out resOr: i7) {
  %0 = comb.and %arg0    : i7
  %1 = comb.and %0, %arg1 : i7
  %2 = comb.or %arg0    : i7
  %3 = comb.or %0, %arg1 : i7
  hw.output %1, %3 : i7, i7
}
```